File: code/snippets/swap-chars/swap.cpp
#include <iostream>
#include <string>
using namespace std;

// strategy: mutate string
// in-place using pointers
void swapChars(string& str, char char1, char char2) {
  for (int i = 0; i < str.length(); i++) {
    str[i] = str[i] == char1 ? 
      char2 : (str[i] == char2 ? char1 : 
        str[i]);
  }
}

int main(int argc, char const *argv[]) {
  // prints @rolling_codes
  string str1 = "@rollinc_godes";
  swapChars(str1, 'c', 'g');
  cout << str1 << endl;
  // prints 314 in binary
  string str2 = "1111111011000101";
  swapChars(str2, '0', '1');
  cout << str2 << endl;
  return 0;
}

File: code/snippets/swap-chars/swap.swift
// strategy: build new string
// char-by-char
func swapChars(in str: String, swapping char1: Character, with char2: Character) -> String {
  var newStr = ""
  for char in str {
    if char == char1 {
      newStr.append(char2)
    } else if char == char2 {
      newStr.append(char1)
    } else {
      newStr.append(char)
    }
  }
  return newStr
}

// prints @rolling_codes
print(
  swapChars(in: "@roccing_lodes", swapping: "c", with: "l"))
// prints 1618 in binary
print(
  swapChars(in: "1111100110101101", swapping: "0", with: "1"))